pkg/quota/controller/clusterquotamapping/clusterquotamapping.go

```diff
@@ -0,0 +1,607 @@
+				break
+			}
+		}
+	}
+
+	c.clusterQuotaMapper.completeNamespace(namespace)
+	return nil
+}
+
+func (c *ClusterQuotaMappingController) quotaWork() bool {
+	key, quit := c.quotaQueue.Get()
+	if quit {
+		return true
+	}
+	defer c.quotaQueue.Done(key)
+
+	quota, exists, err := c.quotaLister.GetByKey(key.(string))
+	if !exists {
+		c.quotaQueue.Forget(key)
+		return false
+	}
+	if err != nil {
+		utilruntime.HandleError(err)
+		return false
+	}
+
+	err = c.syncQuota(quota.(*quotaapi.ClusterResourceQuota))
+	outOfRetries := c.quotaQueue.NumRequeues(key) > 5
+	switch {
+	case err != nil && outOfRetries:
+		utilruntime.HandleError(err)
+		c.quotaQueue.Forget(key)
+
+	case err != nil && !outOfRetries:
+		c.quotaQueue.AddRateLimited(key)
+
+	default:
+		c.quotaQueue.Forget(key)
+	}
+
+	return false
+}
+
+func (c *ClusterQuotaMappingController) quotaWorker() {
+	for {
+		if quit := c.quotaWork(); quit {
+			return
+		}
+	}
+}
+
+func (c *ClusterQuotaMappingController) namespaceWork() bool {
+	key, quit := c.namespaceQueue.Get()
+	if quit {
+		return true
+	}
+	defer c.namespaceQueue.Done(key)
+
+	namespace, exists, err := c.namespaceLister.GetByKey(key.(string))
+	if !exists {
+		c.namespaceQueue.Forget(key)
+		return false
+	}
+	if err != nil {
+		utilruntime.HandleError(err)
+		return false
+	}
+
+	err = c.syncNamespace(namespace.(*kapi.Namespace))
+	outOfRetries := c.namespaceQueue.NumRequeues(key) > 5
+	switch {
+	case err != nil && outOfRetries:
+		utilruntime.HandleError(err)
+		c.namespaceQueue.Forget(key)
+
+	case err != nil && !outOfRetries:
+		c.namespaceQueue.AddRateLimited(key)
+
+	default:
+		c.namespaceQueue.Forget(key)
+	}
+
+	return false
+}
+
+func (c *ClusterQuotaMappingController) namespaceWorker() {
+	for {
+		if quit := c.namespaceWork(); quit {
+			return
+		}
+	}
+}
+
+func (c *ClusterQuotaMappingController) waitForSyncedStores(ready chan<- struct{}, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+
+	for !c.namespacesSynced() || !c.quotasSynced() {
+		glog.V(4).Infof("Waiting for the caches to sync before starting the quota mapping controller workers")
+		select {
+		case <-time.After(100 * time.Millisecond):
+		case <-stopCh:
+			return
+		}
+	}
+	close(ready)
+}
+
+func (c *ClusterQuotaMappingController) deleteNamespace(obj interface{}) {
+	ns, ok1 := obj.(*kapi.Namespace)
+	if !ok1 {
+		tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			utilruntime.HandleError(fmt.Errorf("couldn't get object from tombstone %v", obj))
+			return
+		}
+		ns, ok = tombstone.Obj.(*kapi.Namespace)
+		if !ok {
+			utilruntime.HandleError(fmt.Errorf("tombstone contained object that is not a Namespace %v", obj))
+			return
+		}
+	}
+
+	c.clusterQuotaMapper.removeNamespace(ns.Name)
+}
+
+func (c *ClusterQuotaMappingController) addNamespace(cur interface{}) {
+	c.enqueueNamespace(cur)
+}
+func (c *ClusterQuotaMappingController) updateNamespace(old, cur interface{}) {
+	c.enqueueNamespace(cur)
```

Comment:
> Do you need to enqueue in case there are no label changes for the namespace? Same question about quota without selector changes in its own update handler.

The enqueue method is the spot that does that check since it is common to both the add and update paths.